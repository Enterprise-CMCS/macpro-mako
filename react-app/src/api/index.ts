export * from "./useSearch";
export * from "./useGetForm";
export * from "./useGetItem";
export * from "./useGetUser";
export * from "./getAttachmentUrl";
export * from "./useGetPackageActions";
export * from "./useGetCPOCs";
export * from "./useGetSystemNotifs";
export * from "./useGetTypes";
export * from "./amplifyConfig";
export * from "./itemExists";
export * from "./useGetCounties";
