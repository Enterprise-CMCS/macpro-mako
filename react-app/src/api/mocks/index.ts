export * as mockSubmit from "./submit";
export * as mockTypes from "./types";
