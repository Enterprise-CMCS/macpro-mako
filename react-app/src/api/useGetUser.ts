import { useQuery } from "@tanstack/react-query";
import { Auth } from "aws-amplify";
import { CognitoUserAttributes } from "shared-types";
import { isCmsUser } from "shared-utils";

export type OneMacUser = {
  isCms?: boolean;
  user: CognitoUserAttributes | null;
  counties?: { label: string; value: string }[];
};

export const getUser = async (): Promise<OneMacUser> => {
  try {
    const currentAuthenticatedUser = await Auth.currentAuthenticatedUser();

    if (!currentAuthenticatedUser) {
      return { user: null } satisfies OneMacUser;
    }
    const userAttributesArray = (await Auth.userAttributes(currentAuthenticatedUser)) || [];

    // Set object up with key/values from attributes array
    const userAttributesObj = userAttributesArray.reduce(
      (obj, item) =>
        item?.Name && item?.Value
          ? {
              ...obj,
              [item.Name]: item.Value,
            }
          : obj,
      {} as CognitoUserAttributes,
    );
    // Manual additions and normalizations
    userAttributesObj["custom:cms-roles"] = userAttributesObj["custom:cms-roles"] || "";

    userAttributesObj.username = currentAuthenticatedUser.username || "";

    return {
      user: userAttributesObj,
      isCms: isCmsUser(userAttributesObj),
    } satisfies OneMacUser;
  } catch (e) {
    console.log({ e });
    return { user: null } satisfies OneMacUser;
  }
};

export const useGetUser = () =>
  useQuery({
    queryKey: ["user"],
    queryFn: () => getUser(),
  });
