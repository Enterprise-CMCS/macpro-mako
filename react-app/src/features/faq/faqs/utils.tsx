import { cn } from "@/utils";

export interface Template {
  title: string;
  text?: string;
  href: string;
  subtext?: string[];
}

export const helpDeskContact = {
  email: "OneMAC_Helpdesk@cms.hhs.gov",
  phone: "(833) 228-2540",
};

export const slugify = (text: string) =>
  text
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/(^-|-$)/g, "")
    .substring(0, 40);

export const handleSupportLinkClick = (type: string) => (e: React.MouseEvent<HTMLElement>) => {
  const text = e.currentTarget.textContent?.trim() || "unknown";
  window.gtag("event", `support_click_${type}_${slugify(text)}`, {
    event_category: "Support",
    event_label: text,
  });
};

export const PdfLink = ({
  href,
  title,
  text,
  label,
  className = "underline hover:no-underline",
}: {
  href: string;
  title: string;
  text?: string;
  label: string;
  className?: string;
}) => (
  <a
    className={cn("text-primary", className)}
    href={href}
    download
    rel="noopener noreferrer"
    target="_blank"
    onClick={handleSupportLinkClick(label)}
  >
    {title}
    {text && `: ${text}`}
  </a>
);

export const PdfList = ({
  list,
  label,
  ulClassName = "list-disc pl-6 space-y-2",
}: {
  list: Template[];
  label: string;
  ulClassName?: string;
}) => (
  <ul className={ulClassName} role="list">
    {list.map((pdf) => (
      <li key={pdf.title}>
        <PdfLink href={pdf.href} title={pdf.title} text={pdf.text} label={label} className="" />
        {pdf.subtext && (
          <ul className="list-disc pl-7 space-y-1" role="list">
            {pdf.subtext.map((sub, index) => (
              <li key={index} className="text-sm text-gray-600">
                {sub}
              </li>
            ))}
          </ul>
        )}
      </li>
    ))}
  </ul>
);
