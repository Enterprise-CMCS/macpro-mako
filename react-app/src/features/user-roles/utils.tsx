export type StatusType = "active" | "pending" | "denied" | "revoked";
export type UserRoleType = {
  id: string;
  email: string;
  fullName: string;
  role: string;
  territory: string;
  doneByEmail: string;
  doneByName: string;
  lastModifiedDate: number;
  status: StatusType;
  eventType: string;
  group?: string;
  division?: string;
};

export const initSortUserData = (userData: UserRoleType[]) => {
  if (!userData.length) return [];
  // separate pending / other
  const pendingRoles = userData.filter((x: UserRoleType) => x.status === "pending");
  const remainingRoles = userData.filter((x: UserRoleType) => x.status !== "pending");

  const compare = (a: UserRoleType, b: UserRoleType) => {
    const nameA = a.fullName.toLocaleLowerCase();
    const nameB = b.fullName.toLocaleLowerCase();

    if (nameA < nameB) return -1;
    if (nameA > nameB) return 1;
    return 0;
  };
  const sorted = pendingRoles.sort(compare).concat(remainingRoles.sort(compare));

  return sorted;
};

export const sortUserData = (
  sortByKey: keyof UserRoleType,
  direction: boolean,
  data: UserRoleType[],
) => {
  if (!data.length) return [];

  // when direction is true, that means we are descending
  const [last, first] = direction ? [-1, 1] : [1, -1];
  const sortStatus = (a: UserRoleType, b: UserRoleType) => {
    switch (a.status) {
      case "pending":
        return first;
      case "active":
        return b.status === "pending" ? last : first;
      case "denied":
        return b.status === "revoked" ? first : last;
      case "revoked":
        return last;
      default:
        return last;
    }
  };
  const mainSort = (a: UserRoleType, b: UserRoleType) =>
    a[sortByKey] < b[sortByKey] ? first : last;

  if (sortByKey === "status") return data.sort((a, b) => sortStatus(a, b));

  return data.sort((a, b) => mainSort(a, b));
};
