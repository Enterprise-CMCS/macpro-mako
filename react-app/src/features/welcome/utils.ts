import { QueryClient } from "@tanstack/react-query";
import { redirect } from "react-router";

import { getUser } from "@/api";

export const loader = (queryClient: QueryClient) => {
  return async () => {
    const queryString = window.location.search;
    // Parse the query string to get URL parameters
    const queryParams = new URLSearchParams(queryString);
    // Access specific parameters
    const errorDescription = queryParams.get("error_description");
    const error = queryParams.get("error");
    if (errorDescription || error) {
      console.error("Authentication Error:", { errorDescription, error });
      return { error };
    }

    // check user query has been initialized
    if (!queryClient.getQueryData(["user"])) {
      const userFetch = await queryClient.fetchQuery({
        queryKey: ["user"],
        queryFn: () => getUser(),
      });
      const isUserLoggedIn =
        !["/login", "/faq"].includes(window.location.pathname) && !userFetch?.user;

      if (isUserLoggedIn) {
        return redirect("/login");
      }
    }

    return queryClient.getQueryData(["user"]);
  };
};
