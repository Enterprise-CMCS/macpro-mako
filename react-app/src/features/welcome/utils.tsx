import { QueryClient } from "@tanstack/react-query";
import { Navigate, redirect } from "react-router";

import { getUser, OneMacUser } from "@/api";

export const loader = (queryClient: QueryClient) => {
  return async () => {
    const queryString = window.location.search;
    // Parse the query string to get URL parameters
    const queryParams = new URLSearchParams(queryString);
    // Access specific parameters
    const errorDescription = queryParams.get("error_description");
    const error = queryParams.get("error");
    if (errorDescription || error) {
      console.error("Authentication Error:", { errorDescription, error });
      return { error };
    }

    // check user query has been initialized
    if (!queryClient.getQueryData(["user"])) {
      const userFetch = await queryClient.fetchQuery({
        queryKey: ["user"],
        queryFn: () => getUser(),
      });
      return !userFetch?.user && redirect("/login");
    }

    // check user is logged in
    const loginRedirect =
      !["/login", "/faq", "/support"].includes(window.location.pathname) &&
      !queryClient.getQueryData<OneMacUser>(["user"])?.user;
    if (loginRedirect) {
      return <Navigate to="/login" />;
    }

    return queryClient.getQueryData(["user"]);
  };
};
