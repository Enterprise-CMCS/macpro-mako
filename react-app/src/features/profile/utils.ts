import { UserRole } from "shared-types/events/legacy-user";

import { StateAccess } from "@/api";
import { convertStateAbbrToFullName } from "@/utils";

export const userRoleMap = {
  defaultcmsuser: "CMS Read-only User",
  cmsroleapprover: "CMS Role Approver",
  cmsreviewer: "CMS Read-only User",
  statesystemadmin: "State System Admin",
  helpdesk: "Helpdesk",
  statesubmitter: "State Submitter",
  systemadmin: "System Admin",
};

export const stateAccessRoles: UserRole[] = [
  "statesubmitter",
  "statesystemadmin",
  "cmsroleapprover",
  "systemadmin",
];

export const orderStateAccess = (accesses: StateAccess[]) => {
  if (!accesses || !accesses.length) return;
  // sort revoked states seprately and add to
  const activeStates = accesses.filter((x: StateAccess) => x.status != "revoked");
  const revokedStates = accesses.filter((x: StateAccess) => x.status == "revoked");

  const compare = (a: StateAccess, b: StateAccess) => {
    const stateA = convertStateAbbrToFullName(a.territory);
    const stateB = convertStateAbbrToFullName(b.territory);

    if (stateA < stateB) return -1;
    if (stateA > stateB) return 1;
    return 0;
  };

  const sorted = activeStates.sort(compare).concat(revokedStates.sort(compare));

  return sorted;
};

// if user has no active roles, show pending state(s)
// show state(s) for latest active role
export const filterStateAccess = (userDetails, userProfile) => {
  if (!userProfile?.stateAccess || userProfile.stateAccess.length < 1) return [];
  return userDetails?.role
    ? userProfile.stateAccess.filter(
        (access: StateAccess) => access.role === userDetails.role && access.territory !== "ZZ",
      )
    : userProfile.stateAccess.filter((access: StateAccess) => access.territory !== "ZZ");
};
