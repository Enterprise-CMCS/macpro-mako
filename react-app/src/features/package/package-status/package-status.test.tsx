import { describe, expect, it, vi } from "vitest";

import * as api from "@/api/useGetUser";
import { UseQueryResult } from "@tanstack/react-query";
import { PackageStatusCard } from ".";
import { renderWithQueryClient } from "@/utils/test-helpers";
import { mockUseGetUser, TEST_PACKAGE_STATUS_ITEM } from "mocks";
import { OneMacUser } from "@/api";
describe("Package Status test", () => {
  vi.spyOn(api, "useGetUser").mockImplementation(() => {
    const response = mockUseGetUser();
    response.data.isCms = true;
    return response as UseQueryResult<OneMacUser, unknown>;
  });
  it("The package status component is called successfully", async () => {
    const { asFragment } = renderWithQueryClient(
      <PackageStatusCard data={TEST_PACKAGE_STATUS_ITEM} />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
