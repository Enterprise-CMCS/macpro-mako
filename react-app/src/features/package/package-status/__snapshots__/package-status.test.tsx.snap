// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Package Status test > The package status component is called successfully 1`] = `
<DocumentFragment>
  <div
    class="my-6 sticky top-12"
  >
    <div
      class="h-2 shadow-lg"
      style="border-radius: 3px 3px 0px 0px; height: 8px;"
    />
    <div
      class="border border-t-0 rounded-b-sm border-slate-300"
    >
      <div
        class="p-4 py-1 w-80 min-h-36"
      >
        <h2>
          Status
        </h2>
        <div
          class="my-3 max-w-2xl font-bold text-xl"
        >
          <div />
          <div
            class="flex mt-1 flex-col gap-1 items-start"
          >
            <div
              class="flex flex-row gap-1"
            >
              <p
                class="text-xs font-bold opacity-80"
              >
                ·
              </p>
              <p
                class="text-xs opacity-80"
              >
                Withdraw Formal RAI Response - Enabled
              </p>
            </div>
            <div
              class="flex flex-row gap-1"
            >
              <p
                class="text-xs font-bold opacity-80"
              >
                ·
              </p>
              <p
                class="text-xs opacity-80"
              >
                2nd Clock
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  ,
</DocumentFragment>
`;
